pkg/services/ngalert/provisioning/notification_policies.go

```diff
@@ -154,3 +173,117 @@ func (nps *NotificationPolicyService) ensureDefaultReceiverExists(cfg *definitio
 	nps.log.Error("Grafana Alerting has been configured with a default configuration that is internally inconsistent! The default configuration's notification policy must have a corresponding receiver.")
 	return fmt.Errorf("inconsistent default configuration")
 }
+
+func calculateRouteFingerprint(route definitions.Route) (string, error) {
+	sum := fnv.New64a()
+	err := writeToHash(sum, &route)
+	if err != nil {
+		return "", fmt.Errorf("failed to calculate fingerprint: %w", err)
+	}
+	return fmt.Sprintf("%016x", sum.Sum64()), nil
+}
+
+func writeToHash(sum hash.Hash, r *definitions.Route) error {
+	writeBytes := func(b []byte) {
+		_, _ = sum.Write(b)
+		// add a byte sequence that cannot happen in UTF-8 strings.
+		_, _ = sum.Write([]byte{255})
+	}
+	writeString := func(s string) {
+		if len(s) == 0 {
+			writeBytes(nil)
+			return
+		}
+		// #nosec G103
+		// avoid allocation when converting string to byte slice
+		writeBytes(unsafe.Slice(unsafe.StringData(s), len(s)))
+	}
+
+	// this temp slice is used to convert ints to bytes.
+	tmp := make([]byte, 8)
+	writeInt := func(u int64) {
+		binary.LittleEndian.PutUint64(tmp, uint64(u))
+		writeBytes(tmp)
+	}
+	writeBool := func(b bool) {
+		if b {
+			writeInt(1)
+		} else {
+			writeInt(0)
+		}
+	}
+	writeDuration := func(d *model.Duration) {
+		if d != nil {
+			writeInt(int64(*d))
+		} else {
+			writeInt(0)
+		}
+	}
+
+	writeString(r.Receiver)
+	for _, s := range r.GroupByStr {
+		writeString(s)
+	}
+	for _, labelName := range r.GroupBy {
+		writeString(string(labelName))
+	}
+	writeBool(r.GroupByAll)
+	if len(r.Match) > 0 {
+		keys := maps.Keys(r.Match)
+		slices.Sort(keys)
+		for _, key := range keys {
+			writeString(key)
+			writeString(r.Match[key])
+		}
+	}
+	if len(r.MatchRE) > 0 {
+		keys := maps.Keys(r.MatchRE)
+		slices.Sort(keys)
+		for _, key := range keys {
+			writeString(key)
+			str, err := r.MatchRE[key].MarshalJSON()
+			if err != nil {
+				return err
```

Comment:
Let's avoid the error in the signature altogether using the strategy from receiver fingerprints:

```go
if err != nil {
    f.writeString(fmt.Sprintf("%+v", v))
}
```